pos-print: add async receipt printing

tsbonev opened the pull request:
Added a receipt repository in mongodb and an async printing service, utilizing the mongo capped collection and a background thread to print receipts after they have been persisted.

tsbonev committed:
pos-print: add async receipt printing

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/PosPrintService.java:
```diff
@@ -39,6 +40,10 @@ public static void main(String[] args) {
       new PersistentModule(client, commandCLI.dbName())
     );
 
+    BackgroundReceiptPrinter backgroundPrinter = injector.getInstance(BackgroundReceiptPrinter.class);
```
You can use guava services about this as it will enforce lifecycle (start/stop) of the printing service.

tsbonev replied:
Implemented Guava services for the background thread.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/db/BackgroundReceiptPrinter.kt:
```diff
@@ -0,0 +1,12 @@
+ */
+interface BackgroundReceiptPrinter {
+  fun run()
+  fun printReceipts(collection: MongoCollection<Document>)
```
BackgroundReceiptPrinter interface gets coupled to mongodb which doesn't sound good. It would be better db to be passed to the constructor.

tsbonev replied:
I will implement an interface to abstract the receipt queue from the background thread and that should clear up these dependencies.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/db/BackgroundReceiptPrinterService.kt:
```diff
@@ -0,0 +1,99 @@
+
+  override fun run() {
+    logger.info("Starting background printing service")
+    repository.getCappedCollection().drop()
```
This drop will cause issues If app is re-started and some receipts are not printed.

tsbonev replied:
All of the receipts are stored in another collection before being passed to the capped one, so none would be lost, the drop is there because a restart would print all of the contained receipts and removing a single receipt from the capped collection isn't allowed.

tsbonev replied:
The use of a capped collection here seemed to mimic a normal Java queue as it would need to be dropped at the start and end of the program. On top of that, the mongo tailable cursor was not behaving properly so I switched the implementation for a plain in memory queue that block when asked for the next element and the functionality has remained the same with the bonus of loosing Mongo's capped collection restrictions.

mgenov replied:
Using of in-memory queue is not a good idea as it doesn't allow multiple instances of the pos-print service to be used which removes the redundancy.

Take a look into this one:
http://learnmongodbthehardway.com/schema/queues/

> A queue in the setting of MongoDB, means a collection where inserted documents are kept in an order, either by the creation data of the document or by a priority assigned to the document. Multiple publishers can be inserting new documents into the queue at the same time, but each document can only be delivered to a single listener.


e.g this is what is expected from search service.

tsbonev replied:
Multiple instances would all have their own in-memory blocking queue, why would there be a need for a shared queue between the instances?

mgenov replied:
Because both could read the same message and try to print it.

mlesikov replied:
>Using of in-memory queue is not a good idea as it doesn't allow multiple instances of the pos-print service to be used which removes the redundancy.

Currently we cannot use many instances of the pos-print service because we will have and more that one workers, so many pos-print services can write in the same queue receipts for a a same pos printer, and many workers will get them and sen them concurrently to the same printer.

If we would like to use many instances, we should first separate the app to two different apss one for reading and adding receipts to the queue and one that reads the data in the queue.

Other solution could be creating separate queue for each pos-printer. 

Using the CappedCollections required the both operations : put in the collection and put in the capped collection to be executed in a transaction(I'm not a hundred 100% sure)

mgenov replied:
Yes, the printing is blocking it for now. Ok, keep it with the in-memory implementation.

mgenov replied:
https://github.com/clouway/pos-print/issues/19

mgenov replied:
Queue separation per IP my do it actually.

mlesikov replied:
Also with this in memory queue we can implement logic on the app start, that checks for a not printed receipts and then schedule them in the queue, like that if we restart the app it will load the not printed receipts and trey to print them

mgenov replied:
>Using the CappedCollections required the both operations : put in the collection and put in the capped collection to be executed in a transaction(I'm not a hundred 100% sure)

Yes, it's not atomic now. I suppose it would be better single not capped collection to be used instead as till findAndModify may update it in atomic manner.

> Other solution could be creating separate queue for each pos-printer.

Instances have to take free printer and use it ... this adds a lot of complexity and this is why it would be better to be in-memory. 


As final: the in-memory solution sounds reasonable for know. If we encounter issues then we will may consider implementing another options.

mlesikov replied:
I agree with you. At this time the app works syncronous and there a just a few problems.
I think that we should provide and a rest services, to serve the last recepts and a rest service to schedule already saved receipt for printing again(failed receipts). Like that we could provide simple table with that contains last 50 receipts and show button for rescheduling of the failed.

mgenov commented on pos-print/src/test/java/com/clouway/pos/print/adapter/http/BackgroundReceiptPrinterServiceTest.kt:
```diff
@@ -0,0 +1,126 @@
+      oneOf(notifier).notifyPrinted(acceptedPrintingResponse)
+    }
+
+    collection.insertOne(receiptDoc)
```
Looks really strange to be tested like that as the test knows the structure of the receipt in the queue. The test shouldn't know this and should deal with Receipt object or ReceiptScheduledForPrinting.

tsbonev replied:
Separated mongo logic from the background thread's dependencies.

mgenov commented on pos-print/src/test/java/com/clouway/pos/print/persistent/PersistentReceiptRepositoryTest.kt:
```diff
@@ -0,0 +1,140 @@
+  fun requeueingQueuedReceiptThrowsException() {
+    repository.queue(anyReceipt, sourceIp, isFiscal)
+
+    Thread.sleep(500)
```
Sleep is not a good idea in tests.

tsbonev replied:
Removed thread sleeps from tests.

mlesikov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/db/BackgroundReceiptPrinterService.kt:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  @Suppress("UNCHECKED_CAST")
+  private fun adaptReceipt(receiptDoc: Document): Receipt {
```
you should work with the documents only in the persistence adapter.
You can have PrinterQueue interfece 
```kotlin 
interface PrinterQueue {
    fun first():Receipt
    fun add(receipt: Receipt)
    fun clear()
}
```
and you should work with it here in the logic

tsbonev replied:
Implemented a PrintQueue interface that provides methods to add a receipt, check for a next receipt, get the next receipt and clear itself.

tsbonev commented:
PR is up to date with requested changes.

mgenov commented on pos-print/src/test/java/com/clouway/pos/print/core/BackgroundReceiptPrintWorkerTest.kt:
```diff
@@ -0,0 +1,168 @@
+  var cleaner = DatastoreCleaner(dataStoreRule.db())
+
+
+  private fun Mockery.expecting(block: Expectations.() -> Unit) {
```
Private function should be last in the file.

tsbonev replied:
Moved private functions to the bottom of the file.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/core/PrintNotifier.kt:
```diff
@@ -0,0 +1,16 @@
+ *
+ * @author Tsvetozar Bonev (tsvetozar.bonev@clouway.com)
+ */
+interface PrintNotifier {
```
PrintNotifier doesn't sound good as interface. PrintingListener sounds better as name.

```kotlin
onReceiptPrinted(receipt: Receipt, printStatus: PrintStatus)
```

tsbonev replied:
Changed interface name to PrintingListener and implemented the suggested method signature.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/core/PrintQueue.kt:
```diff
@@ -0,0 +1,31 @@
+   * Checks if there is a next receipt, blocks until a next
+   * one is found.
+   */
+  fun hasNext(): Boolean
```
I'm not sure that you need hasNext. next() could block.

For example take a look into: https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/ArrayBlockingQueue.html#poll--

tsbonev replied:
Implemented blocking via the ArrayBlockingQueue take() method.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/PosPrintService.java:
```diff
@@ -39,9 +40,13 @@ public static void main(String[] args) {
     HttpBackend backend = new HttpBackend(commandCLI.httpPort(), injector);
     backend.start();
 
+    ReceiptPrintWorker backgroundPrinter = injector.getInstance(ReceiptPrintWorker.class);
```
ReceiptPrintingService sounds better then worker.

tsbonev replied:
Renamed to ReceiptPrintingService.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/PosPrintService.java:
```diff
@@ -39,9 +40,13 @@ public static void main(String[] args) {
+
     System.out.printf("POS Print Service is up and running on port: %d\n", commandCLI.httpPort());
 
     Runtime.getRuntime().addShutdownHook(new Thread(() -> {
```
You have to shutdown printing service when termination signal is received.

tsbonev replied:
awaitTerminated() won't return, is it necessary for it to be called or should I just leave it as stopAsync()?

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/http/PrintServiceV2.kt:
```diff
@@ -0,0 +1,54 @@
+      Reply.with(receiptStatus).ok()
+    }catch (ex: ReceiptNotInQueueException){
+      logger.error("Receipt not found")
+      Reply.saying<Int>().notFound()
```
```saying<Int>``` is strange. You can use Any or Unit I suppose.

tsbonev replied:
Changed to be saying<Unit>.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/core/BackgroundReceiptPrintWorker.kt:
```diff
@@ -0,0 +1,78 @@
+  override fun printReceipts() {
+    while (queue.hasNext()) {
+      val receiptWithSource = queue.next()
+      println(receiptWithSource)
```
logger could be used instead of println

tsbonev replied:
Removed forgotten testing printlns from code.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/core/BackgroundReceiptPrintWorker.kt:
```diff
@@ -0,0 +1,78 @@
+        if (printResponse.warnings.contains(Status.FISCAL_RECEIPT_IS_OPEN)
+          || printResponse.warnings.contains(Status.NON_FISCAL_RECEIPT_IS_OPEN)) {
+          logger.info("Receipt printing accepted")
+          repository.finishPrinting(receipt.receiptId)
```
The repository could be responsible for the notification. This is going to remove notifier/listener dependency from here.

tsbonev replied:
I don't think the repository should be responsible for the notification, as the listener handles a PrintReceiptResponse which the repository otherwise doesn't have to know about.

mgenov replied:
You are partially right but:

Here you have 4 dependencies:
```kotlin
class BackgroundReceiptPrintWorker @Inject constructor(private var repository: ReceiptRepository,
                                                       private var factory: PrinterFactory,
                                                       private var notifier: PrintNotifier,
                                                       private var queue: PrintQueue)
```
which causes tests to become bloating with a lot of code due the number of dependencies. This is a smell and need to be considered split of them.

tsbonev replied:
Should I make the repository depend on the listener interface or should I scrap it and write a hard-coded variant?

mgenov replied:
Could you elaborate "write a hard-coded variant?"

Some snippet will be helpful.

tsbonev replied:
Currently I only have one implementation of the listener that waits for a second and prints to the console which I used for testing.

```java
class ConsolePrintingListener : PrintingListener {
  override fun notifyPrinted(printResponse: PrintReceiptResponse) {
    Thread.sleep(1000)
    println(printResponse)
  }
}
```

My question is if I should pass the repository this interface as a dependency, or if I should have something like a logger that logs when finishReceipt or failReceipt are called.

mgenov replied:
Thread.sleep in testing is not a good idea.

Actually both the Repository and ConsolePrintingListener could implement PrintingListener interface. The persistence will record the result whether the console will print it to the console and this class will replace 2 dependencies with one.

tsbonev replied:
Something like this?

```java
class PersistentReceiptRepository @Inject constructor(private val database: Provider<MongoDatabase>)
  : ReceiptRepository, PrintingListener {
```

```kotlin
override fun notifyPrinted(printResponse: PrintReceiptResponse) {
    val logger = LoggerFactory.getLogger(PrintingListener::class.java)

    logger.info("Receipt was printed")
    logger.info(printResponse.warnings.toString())
  }
```

mgenov replied:
Yes, just change the name from notifyPrinted to something like: onReceiptPrinted 

as listener is a class which receive events and not direct orders like "notifyPrinted".

tsbonev replied:
Okay, understood.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/core/InMemoryPrintQueue.kt:
```diff
@@ -0,0 +1,29 @@
+ */
+class InMemoryPrintQueue : PrintQueue {
+
+  private val queue = BlockingArrayQueue<PrintReceiptRequest>()
```
ArrayBlockingQueue is java equivalent which is better to be used as this one is related to jetty.

tsbonev replied:
My mistake, changed to java blocking queue.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/http/PrintServiceV2.kt:
```diff
@@ -0,0 +1,54 @@
+      queue.queue(PrintReceiptRequest(dto.receipt, dto.sourceIp, dto.fiscal))
+
+      logger.info("Receipt queued for printing")
+      Reply.with(receiptId).ok()
```
status 202 is better to be returned as it will indicate that background processing will be performed.

https://httpstatuses.com/202

tsbonev replied:
Changed to return status 202.

mgenov commented on pos-print/src/main/java/com/clouway/pos/print/core/ReceiptPrintWorker.kt:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @author Tsvetozar Bonev (tsvetozar.bonev@clouway.com)
+ */
+interface ReceiptPrintWorker {
```
Not sure that you need this interface at all.

tsbonev replied:
Agreed, but if I remove this interface from the background thread and leave it as a plain AbstractExecutionThreadSerivce, how would I go about binding it's injections with Guice?

mgenov replied:
Do you need such binding at all?

mgenov replied:
btw, you can bind and classes too.

tsbonev replied:
I do not, my mistake, will remove the interface.

tsbonev commented:
Updated Swagger with new v2 api endpoints

mlesikov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/db/PersistentReceiptRepository.kt:
```diff
@@ -0,0 +1,136 @@
+  override fun getStatus(receiptId: String): PrintStatus {
+    val receiptDoc = receipts().find(
+      eq("receiptId", receiptId)).firstOrNull()
+      ?: throw ReceiptNotInQueueException()
```
why ReceiptNotInQueueException ? 
I think ReceiptNotFoundException  fits better, actually we check the status.
We can have a receipt with status printed, that will not be in the queue for printing, that's why the name of the exception sounds confusing to me.

tsbonev replied:
Renamed exceptions to ReceiptNotFoundException and ReceiptAlreadySavedExceptions for clarity.

mlesikov commented on pos-print/src/main/java/com/clouway/pos/print/core/BackgroundReceiptPrintingService.kt:
```diff
@@ -0,0 +1,76 @@
+      }catch (ex: IOException){
+        logger.warn("Printer threw IO Exception")
+        repository.failPrinting(receipt.receiptId)
+      }finally {
```
hm, maybe we can handle and log if there any other unexpected exception, otherwise the background process will fail and the app should be restarted.

tsbonev replied:
Added a catch for unexpected exceptions and logged their message.

mlesikov commented on pos-print/src/main/java/com/clouway/pos/print/core/PrintingListener.kt:
```diff
@@ -0,0 +1,16 @@
+   *
+   * @param printResponse the response to send
+   */
+  fun onReceiptPrinted(receipt: Receipt, printStatus: PrintStatus)
```
hm this interface doesn't look required for any functionality

mgenov replied:
I disagree.

mlesikov replied:
Where this interface is used in the code, I couldn't find a place

tsbonev replied:
It's implemented in the Receipt Repository to avoid bunching up dependencies in the background service.

mlesikov replied:
the implementation looks like this : 
```
override fun onReceiptPrinted(receipt: Receipt, printStatus: PrintStatus) {
    val logger = LoggerFactory.getLogger(PrintingListener::class.java)
     logger.info("Receipt was processed")
    logger.info(receipt.receiptId)
    logger.info(printStatus.name)
  }
```
it is placed in the PersistentReceiptRepository, also this method is called from the same class and it is not used anywhere else in the app, that's why I think at this time is not used at all.
In this version of the PR, this interface is not properly used and the code will work and without it.

tsbonev replied:
I implemented the interface in a separate class and injected it into the repository to avoid it having to implement both interfaces, but I disagree that the interface serves no purpose as the current logging notification will not be the only effect that will happen if we want to notify the user of the status in the future and thus an interface needs to exist as a grounds for that future feature.

tsbonev commented:
PR is up to date with requested changes.

tsbonev commented:
Wrote a fake printer that writes to a file and waits five second and tested the queue, it works as expected, a request containing 1000 receipts gets processed in around five seconds and the receipts begin to get printed in the background, the endpoint that checks their status returns PRINTING when they have not yet been written to a file and PRINTED when they have.
Making a new request while the queue is printing does not disturb the queue.

Currently, if the request exceeds the printing queue limit of 50 when the queue is filled up it will not take in new requests and most of the 1000 receipts will only be persisted and not queued, should this be changed to allow for the queueing of receipts over the queue limit?

```kotlin
class FakePrinter : ReceiptPrinter {
  override fun printReceipt(receipt: Receipt?): PrintReceiptResponse {
    return PrintReceiptResponse(setOf(Status.NON_FISCAL_RECEIPT_IS_OPEN))
  }

  override fun printFiscalReceipt(receipt: Receipt): PrintReceiptResponse {

    File("$fileLocation/${receipt.receiptId}.txt").writeText(receipt.receiptItems.toString())

    Thread.sleep(5000)
    return PrintReceiptResponse(setOf(Status.NON_FISCAL_RECEIPT_IS_OPEN))
  }

  override fun reportForOperator(operatorId: String?, state: RegisterState?) {}

  override fun reportForPeriod(start: LocalDateTime?, end: LocalDateTime?, periodType: PeriodType?) {}

  override fun close() {}
}
```

tsbonev commented:
Added #66 to clarify changes to persistence in the PR.
Implementation is WIP.

tsbonev commented:
Added endpoint to retrieve the last receipts along with their request id and printing status by their creation date and with a limit argument.
Added endpoint to requeue a receipt by request id.

mlesikov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/http/PrintServiceV2.kt:
```diff
@@ -0,0 +1,97 @@
+      receiptList.add(ReceiptWithStatusDTO(it.value, it.key, Gson().toJson(receipts[it.key]!!)))
+    }
+
+    return Reply.with(receiptList).ok()
```
Reply.with(receiptList).`as`(GsonTransport::class.java).ok()

mlesikov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/http/PrintServiceV2.kt:
```diff
@@ -0,0 +1,97 @@
+    }
+  }
+
+  internal data class ReceiptWithStatusDTO(val status: String, val requestId: String, val receiptJson: String)
```
suggest
```
ReceiptDTO(val id:String, val amount:String, val requestId: String, val status: String, val isFiscal:Boolean, val amount:Double, val referent...)
```

mlesikov commented on pos-print/src/main/java/com/clouway/pos/print/adapter/http/PrintServiceV2.kt:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  internal data class ReceiptWithStatusDTO(val status: String, val requestId: String, val receiptJson: String)
+  internal data class ReceiptDTO(val sourceIp: String = "", val operatorId: String = "", val fiscal: Boolean = false, val receipt: Receipt = Receipt.newReceipt().build())
```
tI think that it is better if this dto is PrintReceiptRequest(...)

it look like you are using the domain object as a dto : 
Receipt = Receipt.newReceipt().build()

you should create separate dto object that is used only for parsing the request , then you will have to make transformation from dto to the domain

tsbonev replied:
I implemented a ReceiptWithStatus object that is now stored in persistence, containing information about it's printing status, fiscal status, request id, source ip and its receipt object.

I don't think using the receipt domain object directly here will cause a problem, as the request contains a fully-fledged receipt object within itself already.

StanimirIliev commented on pos-print/src/main/java/com/clouway/pos/print/adapter/db/PersistentModule.kt:
```diff
@@ -1,5 +1,7 @@
 package com.clouway.pos.print.adapter.db
 
+import com.clouway.pos.print.core.PrintingListener
```
Unused import

tsbonev replied:
Removed the unused import.

StanimirIliev commented on pos-print/src/main/java/com/clouway/pos/print/adapter/db/PersistentReceiptRepository.kt:
```diff
@@ -0,0 +1,196 @@
+    if (receipts()
+        .find(and(
+          eq("receipt.receiptId", receipt.receiptId),
+          eq("isFiscal", receiptRequest.isFiscal))).firstOrNull() != null)
```
Instead of ```.firstOrNull() != null``` you could use ```.any()```

tsbonev replied:
Switched over to using any().

StanimirIliev commented on pos-print/src/main/java/com/clouway/pos/print/adapter/db/PersistentReceiptRepository.kt:
```diff
@@ -0,0 +1,196 @@
+          eq("isFiscal", receiptRequest.isFiscal))).firstOrNull() != null)
+      throw ReceiptAlreadyRegisteredException()
+
+    val requestId = UUID.randomUUID().toString()
```
It is better for me to use an interface for generating ids, this way the testing would be easier and also if you decide to not using ```UUID``` you will make changes only in one place

tsbonev replied:
Created an IdGenerator intreface and injected a SimpleUUIDGenerator to the repository.

StanimirIliev commented on pos-print/src/main/java/com/clouway/pos/print/core/ReceiptPrintingListener.kt:
```diff
@@ -0,0 +1,16 @@
+ */
+class ReceiptPrintingListener : PrintingListener {
+  override fun onPrinted(receipt: Receipt, printStatus: PrintStatus) {
+    val logger = LoggerFactory.getLogger(PrintingListener::class.java)
```
It would be better if you bring the initialization out of the function

tsbonev replied:
Must've overlooked this, moved it into the class.

StanimirIliev commented on pos-print/src/test/java/com/clouway/pos/print/core/BackgroundReceiptPrintingServiceTest.kt:
```diff
@@ -0,0 +1,155 @@
+  companion object {
+    @ClassRule
+    @JvmField
+    val dataStoreRule = DatastoreRule()
```
If you use ```@Rule``` instead of ```@ClassRule``` there is no need to make ```dataStoreRule``` static

tsbonev replied:
Making the annotation @Rule throws a null pointer and the only way I've managed to get this rule to work is by making it a static ClassRule.
